src/main/java/net/minecraftforge/event/entity/living/PotionEvent.java

```diff
@@ -0,0 +1,129 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package net.minecraftforge.event.entity.living;
+
+import javax.annotation.Nullable;
+
+import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.potion.Potion;
+import net.minecraft.potion.PotionEffect;
+
+import net.minecraftforge.fml.common.eventhandler.Cancelable;
+import net.minecraftforge.fml.common.eventhandler.Event.HasResult;
+
+/**
+ * This Event and its subevents gets fired from  {@link EntityLivingBase} on the  {@link MinecraftForge#EVENT_BUS}.<br>
+ */
+public class PotionEvent extends LivingEvent
+{
+    @Nullable
+    protected final PotionEffect effect;
+    
+    public PotionEvent(EntityLivingBase living, PotionEffect eff)
+    {
+        super(living);
+        effect = eff;
+    }
+    /**
+     * Retuns the PotionEffect
+     */
+    @Nullable
+    public PotionEffect getPotionEffect()
+    {
+        return effect;
+    }
+    
+    /**
+     * This Event is fired when a Potion is about to get removed from an Entity
+     * This Event is {@link Cancelable}
+     * This Event does not have a result.
+     */
+    @Cancelable
+    public static class PotionRemoveEvent extends PotionEvent
+    {
+        private final Potion toRemove;
+        
+        public PotionRemoveEvent(EntityLivingBase living, Potion  potion)
+        {
+            super(living, living.getActivePotionMap().get(potion));
+            toRemove = potion;
+        }
+        
+        /**
+         * Retuns the Potion wich is tried to remove from the Entity
+         */
+        public Potion getPotion()
+        {
+            return toRemove;
+        }
+    }
+    
+    /**
+     * This Event is fired to check if a Potion can get applied to an Entity.
+     * This Event is not {@link Cancelable}
+     * This Event has a result {@link HasResult}.
+     * ALLOW will apply this potion effect
+     * DENY will not apply this potion effect
+     * DEFAULT will run vanilla logic to determine if this potion isApplicable
+     */
+    @HasResult
+    public static class PotionApplicableEvent extends PotionEvent
+    {
+        public PotionApplicableEvent(EntityLivingBase living, PotionEffect eff)
+        {
+            super(living, eff);
+        }
+    }
+    
+    /**
+     * This Event is fired when a new Potion is added to the Entity.
+     * This Event is not {@link Cancelable}
+     * This Event does not have a Result.
+     */
+    public static class PotionAddedEvent extends PotionEvent
+    {
+        public PotionAddedEvent(EntityLivingBase living, PotionEffect eff)
```

Comment:
Can you override the getter of potion effect and mark it as nonnull since this is the case here? Same for the applicable event above

Reply:
Yeah this is a good idea

Reply:
Better idea, I make it notnull and pnly nullable for potion remove ?

Reply:
Or make the remove a different subclass